pilot/api/es_data.py

```diff
@@ -0,0 +1,202 @@
+            storages = self.astorages.get(act)
+            if not copytools:
+                logger.warn("No available copytools for activity %s" % act)
+                continue
+            if act in ['pw', 'pls', 'es_events', 'es_failover'] and not storages:
+                # for write activity, if corresponding storages are not defined, should use different activity
+                logger.warn("Failed to find corresponding astorages for writing activity(%s), will try next activity" % act)
+                continue
+
+            storage = storages[0] if storages else None
+            avail_activity = True
+            for name in copytools:
+                try:
+                    if name not in self.copytool_modules:
+                        raise PilotException('passed unknown copytool with name=%s .. skipped' % name)
+                    module = self.copytool_modules[name]['module_name']
+                    logger.info('Trying to use copytool=%s for activity=%s' % (name, act))
+                    copytool = __import__('pilot.copytool.%s' % module, globals(), locals(), [module], -1)
+                except PilotException as e:
+                    errors.append(e)
+                    logger.debug('Error: %s' % e)
+                    continue
+                except Exception as e:
+                    logger.warning('Failed to import copytool module=%s, error=%s' % (module, e))
+                    logger.debug('Error: %s' % e)
+                    continue
+
+                try:
+                    result = self.transfer_files(copytool, files, act, storage, **kwargs)
+                except PilotException, e:
+                    errors.append(e)
+                    logger.debug('Error: %s' % e)
+                except Exception as e:
+                    logger.warning('Failed to transfer files using copytool=%s .. skipped; error=%s' % (copytool, e))
+                    logger.error(traceback.format_exc())
+                    errors.append(e)
+
+                if errors and isinstance(errors[-1], PilotException) and errors[-1].get_error_code() == ErrorCodes.MISSINGOUTPUTFILE:
+                    raise errors[-1]
+
+                if result:
+                    break
+                else:
+                    logger.warn("Failed to transfer files using activity(%s) copytool(%s) with error=%s" % (act, name, errors))
+            if result:
+                break
+            else:
+                logger.warn("Failed to transfer files using activity(%s) with copytools(%s)" % (act, copytools))
+
+        if not avail_activity:
+            raise PilotException('Not available activity with both acopytools and astorages defined')
+        if not result:
+            raise PilotException('Failed to transfer files with activities %s' % (activity))
+
+        return result
+
+
+class StageInESClient(StagingESClient, StageInClient):
+
+    def get_storage_id_and_path_convention(self, storage_token):
```

Comment:
data validation and parsing should be applied from the beginning/the responsibility of an object which is the master/source of data - FileSpec/StorageData

Reply:
ok. I will move this part to filespec.

Reply:
moved it to filespec.